File: 백준/Gold/2589. 보물섬/보물섬.swift
import Foundation

// start point (i, j)
func solve(i: Int, j: Int) -> Int {
    guard board[i][j] == "L"
    else { return 0 }
    
    let dx: [Int] = [0, 0, 1, -1]
    let dy: [Int] = [1, -1, 0, 0]
    
    var result: Int = 0
    var visited: [[Int]] = Array(repeating: 
                                    Array(repeating: Int.max,
                                          count: w),
                                 count: l)
    
    
    var queue: Queue<(Int, Int)> = .init()
    queue.push((i, j))
    visited[i][j] = 0
    
    while !queue.isEmpty() {
        let (x, y) = queue.popLeft()!
        
        for index in (0..<4) {
            let nx = x + dx[index]
            let ny = y + dy[index]
            
            guard nx >= 0, nx < l, ny >= 0, ny < w,
                  board[nx][ny] == "L"
            else { continue }
            
            if visited[nx][ny] == Int.max {
                queue.push((nx, ny))
            }
            
            visited[nx][ny] = min(visited[nx][ny], visited[x][y] + 1)
            result = max(visited[nx][ny], result)
        }
    }
    
    return result
}


let arr: [Int] = readLine()!.split(separator: " ").map({Int($0)!})
let (l, w) = (arr[0], arr[1])
var board: [[String]] = []

(1...l).forEach { _ in
    board.append(readLine()!.map({String($0)}))
}



var ans: Int = 0

for i in (0..<l) {
    for j in (0..<w) {
        ans = max(ans, solve(i: i, j: j))
    }
}

print(ans)


struct Queue<T> {
    // index
    var left: Int = 0
    var right: Int = 1
    var body: [T] = []
    // body에 있는 개수
    var count: Int = 0
    
    mutating func push(_ newElement: T) {
        self.body.append(newElement)
        self.right += 1
        self.count += 1
    }
    
    mutating func popLeft() -> T? {
        guard self.count > 0
        else { return nil }
        self.left += 1
        self.count -= 1
        return self.body[self.left-1]
    }
    
    func isEmpty() -> Bool {
        return self.count == 0
    }
}

File: 백준/Gold/2589. 보물섬/보물섬.py
#2589
import sys
from collections import *
input = sys.stdin.readline
INF = int(1e9)

def solve(x, y):
    global ans
    dx = [1, -1, 0, 0]
    dy = [0, 0 , 1, -1]
    visited = [[INF for _ in range(w)] for _ in range(l)]
    
    if board[x][y] =='W':
        return 

    visited[x][y] = 0
    q = deque()
    q.append((x,y))
    while q:
        x, y = q.popleft()
        
        for i in range(4):
            nx = x + dx[i]
            ny = y + dy[i]
            
            # out of index
            if nx < 0 or nx >= l or ny < 0 or ny >= w:
                continue
            
            # 바다는 방문하지 않음
            if board[nx][ny] == 'W':
                continue
            
            # 처음 방문하는 경우 
            if visited[nx][ny] == INF:
                q.append((nx, ny))
                
            visited[nx][ny] = min(visited[x][y] + 1, visited[nx][ny])
            ans = max(ans, visited[nx][ny])

l, w = map(int, input().split())
board = [list(input().rstrip()) for _ in range(l)]
ans = 0

for i in range(l):
    for j in range(w):
        solve(i, j)
        
print(ans)